cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -72,19 +72,69 @@ class UniqueAction {
 
   Status Reserve(const int64_t length) { return Status::OK(); }
 
-  void ObserveNull() {}
+  Status ObserveNull() { return Status::OK(); }
 
   template <class Index>
-  void ObserveFound(Index index) {}
+  Status ObserveFound(Index index) {
+    return Status::OK();
+  }
 
   template <class Index>
-  void ObserveNotFound(Index index) {}
+  Status ObserveNotFound(Index index) {
+    return Status::OK();
+  }
+
+  Status Flush(Datum* out) { return Status::OK(); }
+  Status FlushFinal(Datum* out) { return Status::OK(); }
+};
+
+// ----------------------------------------------------------------------
+// Count values implementation (see HashKernel for description of methods)
+
+class CountValuesAction {
+ public:
+  CountValuesAction(const std::shared_ptr<DataType>& type, MemoryPool* pool)
+      : count_builder_(pool) {}
+
+  Status Reserve(const int64_t length) {
+    // builder size is independent of input array size.
+    return Status::OK();
+  }
+  Status Reset() {
+    count_builder_.Reset();
+    return Status::OK();
+  }
 
+  // Don't do anything on flush beceause we don't want to finalize the builder
+  // or incur the cost of memory copies.
   Status Flush(Datum* out) { return Status::OK(); }
+
+  Status FlushFinal(Datum* out) {
+    std::shared_ptr<ArrayData> result;
+    RETURN_NOT_OK(count_builder_.FinishInternal(&result));
+    out->value = std::move(result);
```

Comment:
For completeness, you should set the datum type if we ever refactor.

Reply:
Could you clarify?  I might be missing it but I don't see an API to set the type explicitly (instead it looks like it gets derived when value that is assigned in the underlying variant)

Reply:
It's handled by the variant

Reply:
Yes, my bad.